### ISSUE TITLE
How to show dialog on particular different route in server?

### ISSUE THREAD

**Issue by jayesh83** [dkrivoruchko/ScreenStream#147]
Hi @dkrivoruchko 
It's been nice talking to you. You were really helpful and quick in earlier issues I opened #143 . Thanks for that

Now,
I'm trying to implement a remote access control feature in-app, where I use ScreenStream for streaming.

I've created new rout in `data.httpserver.HttpServerRxHandler` file i.e. `uri == HttpServerFiles.ROOT_ADDRESS + HttpServerFiles.ALLOW_REMOTE_ACCESS`

On the hit of this route, I want to show the dialog in-app module somewhere that says if you want to allow remote access or not by this particular client

- on accepting I want to create a WebSocket, and let the client know somehow on which port WebSocket was created so that client can connect to that (as of now defined set of ports, seeing where I can connect)
- on reject just ignore, do nothing

**Question:**
- How to show dialog on that route hit

**What I know/tried?**
- As per my understanding, ScreenStream is using finite automata where we've defined a bunch of states, actions. Mapping each state with actions using the matrix in `StateToEventMatrix ` file.
- I've thought of two solutions to this. 1. LocalBroadcastManager to send broadcast on route hit 2. An interface in RxHandler, which can be implemented and configured to show a dialog in Stream Viewer
- But I think the way ScreenStream is sending events from data to app module is the correct way

_How to particularly go about implementing my requirement, I'm not getting, Can you help me with that?_

**Comment by jayesh83** [comment]
I'll create PR on the successful implementation of Remote Control.

**Comment by dkrivoruchko** [comment]
Hi.
Current version of HttpServerRxHandler has only one way communication: from AppStateMachine to HttpServerRxHandler.
Keep in mind that nothing in `data` module has access to Android UI and because of Android restrictions we cannot show any dialog when app is minimized.
In order to show user a dialog and send back user decision many things need to be implemented/updated.
My first idea is to add some command channel from HttpServerRxHandler to AppService which can show user a notification with questions.
Letting client to know a user choice and WebSocket configuration is a simple one - just response on request with json containing either configuration or deny.
Open questions are WebSocket and remote control itself - I have no experience in this.

Bottom line - if you show me working code for remote control from web browser to android system - I will update ScreeenStream for requesting control permissions and response with WebSocket config (maybe with WebSocket too, if there will be time for this).

**Comment by jayesh83** [comment]
Hi @dkrivoruchko
Coming to the bottom line, remember we've talked on the issue before #143 where I've mentioned that I'm actually making this app to app and not app to browser.

I've working code for remote control in app to app & not browser to app. So sorry for that matter

**Comment by jayesh83** [comment]
Yeah as you talked about this one way communication from AppStateMachine to HttpServerRxHandler and nothing in the data module has access to Android UI, I completely understood that.

But as you said _`add some command channel from HttpServerRxHandler to AppService`_. Can you explain exatly how to go for it?

As I believe I need to update AppStateMachine add state there and update StateToEventMatrix as well.

So can you talk about exactly in what flow I should go?

**Comment by jayesh83** [comment]
@dkrivoruchko Can you please elaborate the of changes I need to make?

**Comment by dkrivoruchko** [comment]
There is no simple answer and I'm unsure if AppStateMachine or StateToEventMatrix requiere any change.
Do you have code for android control in app to app? Can I have a look on part of it? How this apps talk to each other?

**Comment by jayesh83** [comment]
Yeah! It is against NDA as developing for my company but I can give you an overall idea. we can connect over meet

**Comment by jayesh83** [comment]
I'll close the issue then. Feel free to reopen, to start a discussion again
